File: test/replicator.py
import sys
sys.path.append("..")
from SIMDTierra import *

a=Assembler(0)
a.assembleLines("""
Copy1 R0 PC - decr                        #load start
Copy1 loadAddr PC -                       
Copy1 R0 R0 R0                            # template search pattern
Copy1 loadAddr loadAddr - load incr       # load search pattern to data
Copy1 R2 data -                           # end of search pattern
Copy1 R1 PC -                             # find loop
Copy1 loadAddr loadAddr - load incr       # data=soup[loadAddr]
Eq    test R2 data jmpz                   # if not equal, goto find loop
Copy1 PC R1 -                             # goto find loop
Copy1 R2 loadAddr -                       # R2 = end
Copy1 loadAddr R0 -                       # loadAddr = start
Copy1 R1 PC -                             # copy loop
Copy1 loadAddr loadAddr - load incr       # data=soup[loadAddr]
Copy1 storeAddr storeAddr - store incr   # soup[storeAddr]=data
Xor test loadAddr R2 jmpz                       # if PC==R2, goto end
Copy1 PC R1 R1                            # goto copy loop
""")
# add in complement of template pattern
a.soup[a.current()]=instruction("Not1", ~reg("R0"), ~reg("R0"), ~reg("R0"), "jmpz", "full")

assert(a.soup[a.current()]==4294882723)

cpu=CPU()
cpu.registers[reg("PC")]=0
cpu.registers[reg("storeAddr")]=20

for i in range(500):
    if cpu.registers[reg("PC")]>16: break
    cpu.execute()

for i in range(17):
    #print(i,a.soup[i],a.soup[i+20])
    assert(a.soup[i]==a.soup[i+20])

File: test/replicator.cc
#include <gtest/gtest.h>

#include "../SIMDCPU.h"

#include <iostream>

using namespace SIMDCPU;
using namespace std;

struct TestCPU: public CPU, public testing::Test {};

TEST_F(TestCPU, Replicator) {
  soup[0]=Instruction(Copy1, R0, PC, PC, none, decr).asWord(); // load start
  soup[1]=Instruction(Copy1, loadAddr, PC, PC).asWord();
  soup[2]=Instruction(Copy1, R0, R0, R0).asWord(); // template search pattern
  soup[3]=Instruction(Copy1, loadAddr, loadAddr, loadAddr, load, incr).asWord(); // load search pattern to data
  soup[4]=Instruction(Copy1, R2, data, data).asWord(); // end of search pattern
  soup[5]=Instruction(Copy1, R1, PC, PC).asWord(); // find loop
  soup[6]=Instruction(Copy1, loadAddr, loadAddr, loadAddr, load, incr).asWord(); // load search pattern to data  
  soup[7]=Instruction(Eq, test, R2, data, jmpz).asWord(); // if not equal, goto find loop
  soup[8]=Instruction(Copy1, PC, R1,R1).asWord(); // goto find loop
  soup[9]=Instruction(Copy1, R2, loadAddr, loadAddr).asWord(); // R2 = end
  soup[10]=Instruction(Copy1, loadAddr, R0, R0).asWord();
  soup[11]=Instruction(Copy1, R1, PC, PC).asWord(); // copy loop
  soup[12]=Instruction(Copy1, loadAddr, loadAddr, loadAddr, load, incr).asWord();
  soup[13]=Instruction(Copy1, storeAddr, storeAddr, storeAddr, store, incr).asWord();
  soup[14]=Instruction(Xor, test, PC, R2, jmpz).asWord();
  soup[15]=Instruction(Copy1, PC, R1, R1).asWord(); // goto copy loop
  soup[16]=Instruction(Not1, cmpl(R0), cmpl(R0), cmpl(R0), jmpz, full).asWord(); //complementary template at end

  registers[PC]=0;
  registers[storeAddr]=20;
  for (int i=0; i<500 && registers[PC]<17; ++i)
    execute();
  for (int i=0; i<17; ++i)
    {
      //cout<<i<<endl;
      EXPECT_EQ(soup[i], soup[i+20]);
    }
}
